networkx/lazy_imports.py

```diff
@@ -81,6 +81,18 @@ def __dir__():
     return __getattr__, __dir__, list(__all__)
 
 
+class DelayedImportErrorModule(types.ModuleType):
+    def __init__(self, import_name, *args, **kwargs):
+        self.__import_name = import_name
+        super().__init__(*args, **kwargs)
+
+    def __getattr__(self, x):
+        if x.startswith("__"):
+            super().__getattr__(x)
+        else:
+            raise ModuleNotFoundError(f"No module named '{self.__import_name}'")
```

Comment:
This makes the raised error text the same as for regular imports.
That may cause confusion because the exception is raised potentially much later.
Is there a reason to have a short exception message?
I'd prefer a message that indicates that this is an exception caused by code that executed potentially long ago.

Reply:
Yes, this was a change I made on purpose.

Old behavior:

```
def foo(x):
    import scipy.not_installed
    scipy.not_installed(x)
```

This is what we had before lazy loading: it just raises the moment the function is called.  I'm not sure there is a reason to let the user know that we have a deferred import error now: they'll see, just as before, that scipy.x is missing only once the function tries to use it.

Reply:
Actually, before lazy loading it raised on the import statement itself.  Not on the function call.
The traceback nicely explains that the import is what couldn't find the module.

But with lazy loading the traceback doesn't show any import statement
For example, the old behavior:
```
>>> nx.to_numpy_array(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dschult/NXstefanv/networkx/convert_matrix.py", line 1405, in to_numpy_array
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
```
And the new behavior:
```
>>> nx.utils.misc.dict_to_numpy_array({})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dschult/NXstefanv/networkx/utils/misc.py", line 230, in dict_to_numpy_array
    return dict_to_numpy_array2(d, mapping)
  File "/Users/dschult/NXstefanv/networkx/utils/misc.py", line 248, in dict_to_numpy_array2
    a = np.zeros((n, n))
  File "/Users/dschult/NXstefanv/networkx/lazy_imports.py", line 93, in __getattr__
    raise ModuleNotFoundError(f"No module named '{self.__import_name}'")
ModuleNotFoundError: No module named 'numpy'
```
There is no import statement in sight.    I guess it's not a big difference. But the tracebacks for lazy imports feel confusing enough. 

I'd love something like:
```
>>> nx.utils.misc.dict_to_numpy_array({})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dschult/NXstefanv/networkx/utils/misc.py", line 230, in dict_to_numpy_array
    return dict_to_numpy_array2(d, mapping)
  File "/Users/dschult/NXstefanv/networkx/utils/misc.py", line 248, in dict_to_numpy_array2
    a = np.zeros((n, n))
  File "/Users/dschult/NXstefanv/networkx/lazy_imports.py", line 93, in __getattr__
    raise ModuleNotFoundError(f"'import {self.__import_name}' failed")
ModuleNotFoundError: 'import numpy' failed lazily
```

Reply:
I will modify this to display a more informative error.

Reply:
I changed the error message to look like this:

```
This error is lazily reported, having originally occured in
  File /home/stefan/src/networkx/networkx/tests/test_lazy_imports.py, line 11, in test_lazy_import_basics

----> anything_not_real = lazy.lazy_import("anything_not_real")
```

Reply:
Wow -- so you tracked where the lazy_import was called?!?  That's Cool!

I can't see the code though. Did you push?

Reply:
🤦